skorch/tests/callbacks/test_training.py

```diff
@@ -1140,3 +1140,155 @@ def test_pickle_initialized_callback(self, trainendcheckpoint_cls):
+            n_redundant=0,
+            random_state=0,
+        )
+        return X.astype(np.float32), y
+
+    @pytest.fixture
+    def data_fixed(self):
+        return self.generate_data(n_input=10)
+
+    @pytest.fixture(params=[2, 10, 20])
+    def data_parametrized(self, request):
+        return self.generate_data(n_input=request.param)
+
+    def test_shape_set(
+        self, net_cls, module_cls, input_shape_setter_cls, data_parametrized,
+    ):
+        net = net_cls(module_cls, max_epochs=2, callbacks=[
+            input_shape_setter_cls(),
+        ])
+
+        X, y = data_parametrized
+        n_input = X.shape[1]
+        net.fit(X, y)
+
+        assert net.module_.layer.in_features == n_input
+
+    def test_shape_set_using_fn(
+        self, net_cls, module_cls, input_shape_setter_cls, data_parametrized,
+    ):
+        def input_dim_fn(X):
+            return X.shape[1]
+
+        net = net_cls(module_cls, max_epochs=2, callbacks=[
+            input_shape_setter_cls(input_dim_fn=input_dim_fn),
+        ])
+
+        X, y = data_parametrized
+        n_input = X.shape[1]
+        net.fit(X, y)
+
+        assert net.module_.layer.in_features == n_input
+
+    def test_parameter_name(
+        self, net_cls, input_shape_setter_cls, data_parametrized,
+    ):
+        import torch
+
+        class MyModule(torch.nn.Module):
+            def __init__(self, other_input_dim=22):
+                super().__init__()
+                self.layer = torch.nn.Linear(other_input_dim, 2)
+            def forward(self, X):
+                return self.layer(X)
+
+        net = net_cls(MyModule, max_epochs=2, callbacks=[
+            input_shape_setter_cls(param_name='other_input_dim'),
+        ])
+
+        X, y = data_parametrized
+        n_input = X.shape[1]
+        net.fit(X, y)
+
+        assert net.module_.layer.in_features == n_input
+
+    def test_module_name(
+        self, net_cls, module_cls, input_shape_setter_cls, data_parametrized,
+    ):
+        class MyNet(net_cls):
+            def __init__(self, *args, module2, **kwargs):
+                self.module2 = module2
+                super().__init__(*args, **kwargs)
+
+            def initialize_module(self):
+                kwargs = self.get_params_for('module')
+                self.module_ = self.module(**kwargs)
+
+                kwargs = self.get_params_for('module2')
+                self.module2_ = self.module2(**kwargs)
+
+        net = MyNet(
+            module=module_cls,
+            module2=module_cls,
+            max_epochs=2,
+            callbacks=[
+                input_shape_setter_cls(module_name='module'),
+                input_shape_setter_cls(module_name='module2'),
+            ],
+        )
+
+        X, y = data_parametrized
+        n_input = X.shape[1]
+        net.fit(X, y)
+
+        assert net.module_.layer.in_features == n_input
+        assert net.module2_.layer.in_features == n_input
+
+    def test_no_module_reinit_partial_fit(
```

Comment:
Could you please add a very similar test where the input shape is already correct and you test that after `initialize()`, `fit(*data)` doesn't re-initialize?